src/originalhtml.rs

```diff
@@ -0,0 +1,615 @@
+
+use std::future::Future;
+use std::pin::Pin;
+
+use lettre_email::mime::{self, Mime};
+
+use crate::context::Context;
+use crate::error::Result;
+use crate::headerdef::{HeaderDef, HeaderDefMap};
+use crate::message::{Message, MsgId};
+use crate::mimeparser::parse_message_id;
+use crate::simplify::split_lines;
+use mailparse::ParsedContentType;
+use once_cell::sync::Lazy;
+
+impl Message {
+    /// Check if the mime structure of a Message is modified.
+    /// "Modified" here means that some text is cut or the original message
+    /// is in HTML and simplify() may hide some maybe important information.
+    /// The corresponding ffi-function is dc_msg_is_mime_modified().
+    pub fn is_mime_modified(&self) -> bool {
+        self.mime_modified
+    }
+}
+
+/// Type defining a rough mime-type.
+/// This is mainly useful on interating
+/// to decide whether a mime-part has subtypes.
+enum MimeS {
+    Multiple,
+    Single,
+    Message,
+}
+
+/// Function takes a content type from a ParsedMail structure
+/// and checks and returns the rough mime-type.
+async fn get_mimes(ctype: &ParsedContentType) -> MimeS {
+    let mimetype = ctype.mimetype.to_lowercase();
+    if mimetype.starts_with("multipart") && ctype.params.get("boundary").is_some() {
+        MimeS::Multiple
+    } else if mimetype == "message/rfc822" {
+        MimeS::Message
+    } else {
+        MimeS::Single
+    }
+}
+
+// HtmlMsgParser converts a mime-message to HTML.
+#[derive(Debug)]
+pub struct HtmlMsgParser {
+    pub html: String,
+    pub plain: Option<String>,
+    pub format_flowed: bool,
+    pub delsp: bool,
+}
+
+impl HtmlMsgParser {
+    /// Function takes a raw mime-message string,
+    /// searches for the main-text part
+    /// and returns that as parser.html
+    pub async fn from_bytes(context: &Context, rawmime: &[u8]) -> Result<Self> {
+        let mut parser = HtmlMsgParser {
+            html: "".to_string(),
+            plain: None,
+            format_flowed: false,
+            delsp: false,
+        };
+
+        let parsedmail = mailparse::parse_mail(rawmime)?;
+
+        parser.collect_texts_recursive(context, &parsedmail).await?;
+
+        if parser.html.is_empty() {
+            if let Some(plain) = parser.plain.clone() {
+                parser.html = plain_to_html(&plain, parser.format_flowed, parser.delsp).await;
+            }
+        } else {
+            parser.cid_to_data_recursive(context, &parsedmail).await?;
+        }
+
+        Ok(parser)
+    }
+
+    /// Function iterates over all mime-parts
+    /// and searches for text/plain and text/html parts and saves the
+    /// last one found
+    /// in the corresponding structure fields.
+    /// Usually, there is at most one plain-text and one HTML-text part.
+    fn collect_texts_recursive<'a>(
+        &'a mut self,
+        context: &'a Context,
+        mail: &'a mailparse::ParsedMail<'a>,
+    ) -> Pin<Box<dyn Future<Output = Result<bool>> + 'a + Send>> {
```

Comment:
Where the returned result is used? If it's not needed, maybe drop thin Pin/Box/dyn/Future/Send/lifetime whatever and return Self?

Reply:
Ah, this all is because of async recursion. Anyway, can the `bool` be replaced with `Self`, instead of passing a mutable `self` into this function?

Reply:
also here - i just copied from mimeparser.rs - for me, the whole result line looks more like magic :)

Reply:
there is a newer comment below, marking this as resolved.